### ISSUE TITLE
`aria-required-children` should not fail when `role=listbox` element does not have descendant element of `role=option`

### ISSUE THREAD

**Issue by Jia-Yao** [dequelabs/axe-core#2339]
Expectation: While all options in listbox should have `role=option`, it's not necessarily that every descendant element of the listbox has to be an option. The scenario where `role=listbox` element does not have descendant element of `role=option` should not be treated as failure.

Actual: In axe-core version 3.4.1 that shows up as failure.

Motivation: Issue https://github.com/dequelabs/axe-core/issues/874 has regressed.

<pre><code>
axe-core version: 3.4.1
</code></pre>

**Comment by WilcoFiers** [comment]
Hey @Jia-Yao I'm having a hard time reproducing the issue. Can you provide a code snippet?

**Comment by Jia-Yao** [comment]
Hello @WilcoFiers, here is the high level idea of what I'm trying to do: [https://jsfiddle.net/jia_yao/gq4skxhp/78/](https://jsfiddle.net/jia_yao/gq4skxhp/78/). It's similar to the issue I linked in terms of creating `role=option` elements on the fly, and I see `axe-core` is flagging the `aria-required-children` rule.

**Comment by straker** [comment]
@Jia-Yao this issue may have been fixed in the latest develop version of the code. Would you be able to test your use-case with [axe-core@next](https://www.npmjs.com/package/axe-core/v/next) and see if it works now?

**Comment by straker** [comment]
Closing as the following code snippet passes the required-children rule. Required children should now always work if at least one child is the required role.

```html
<div role="listbox" aria-label="options">
  <div role="option">Foo</div>
  <div>Bar</div>
</div>
```

**Comment by sinap-fb** [comment]
@straker I believe this report was suggesting that `aria-required-children` should pass even if it doesn't have a descendant with `role=option`.  In other words, this test case should pass (but I don't believe it currently does):
```
<div role="listbox" aria-label="options">
  <div>No results</div>
</div>
```

**Comment by straker** [comment]
@sinap-fb `aria-required-children` will require that the element contain at least one of the required roles, but doesn't require all roles to be of the required type. So you're example should fail `aria-required-children` since the listbox does not have the required `option` child.

**Comment by sinap-fb** [comment]
@straker If this is intended behavior, how should a combobox containing a query that has no results be structured to avoid failing `aria-required-children`?  Even in the APG's example [here](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/), if you type in a query with no results, the listbox contains no children, which I believe would fail this rule.

**Comment by dbjorge** [comment]
@sinap-fb, I would recommend using an `option` child with `No results` or similar text. You can see some examples of this pattern in [Microsoft Fluent UI's Combobox](https://react.fluentui.dev/?path=/docs/components-combobox--default#filtering) and [Google Material UI's Autocomplete Combo box](https://mui.com/material-ui/react-autocomplete/#combo-box).

I agree with your assessment that the [APG example you linked](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/) doesn't meet the requirement when filtered down to no options. I think this is an inconsistency between the APG example and the ARIA requirements; [the ARIA standard's section on Required Owned Elements](https://www.w3.org/TR/wai-aria-1.2/#mustContain) is clear that there must exist one of the required children (barring `aria-busy` cases that I don't think are applicable here).

I didn't see an existing issue discussing this in the [aria-practices repo](https://github.com/w3c/aria-practices) where the APG examples are developed - I think you should [feel free to file one](https://github.com/w3c/aria-practices/issues)!
